### PR TITLE
Add drop_unused_requests option to exclude unused interactions

### PR DESCRIPTION
**Pull request by ryanseys**
This is a rebase of https://github.com/vcr/vcr/pull/769 so that we can hopefully merge it! :)

### CODE REVIEW

**Comment by umerakram1** [comment]
Review

File: spec/lib/vcr/cassette_spec.rb [hunk 1]
```diff
@@ -636,31 +637,44 @@ def interaction(response_body, request_attributes)
...
 
-            it 'appends new recorded interactions that do not match existing ones' do
-              expect(saved_recorded_interactions.last).to eq(interaction_bar)
+              it 'appends new recorded interactions that do not match existing ones' do
+                expect(saved_recorded_interactions.last).to eq(interaction_bar)
+              end
             end
+            context "with drop_unused_requests option added to all" do
+              let(:drop_unused_requests) { true }
+              it "drops unused requests" do
```
Comment (olleolleolle) [inline spec/lib/vcr/cassette_spec.rb:669]:
```suggestion

              it "drops unused requests" do
```

File: spec/lib/vcr/cassette_spec.rb [hunk 2]
```diff
@@ -636,31 +637,44 @@ def interaction(response_body, request_attributes)
...
+                expect(saved_recorded_interactions).not_to include(interaction_foo_1)
+              end
 
-            it 'appends new recorded interactions that do not match existing ones' do
-              expect(saved_recorded_interactions.last).to eq(interaction_bar)
+              it 'appends new recorded interactions that do not match existing ones' do
+                expect(saved_recorded_interactions.last).to eq(interaction_bar)
+              end
             end
+            context "with drop_unused_requests option added to all" do
```
Comment (olleolleolle) [inline spec/lib/vcr/cassette_spec.rb:667]:
```suggestion

            context "with drop_unused_requests option added to all" do
```

File: spec/lib/vcr/cassette_spec.rb [hunk 3]
```diff
@@ -611,7 +611,8 @@ def stub_old_interactions(interactions)
 
     [:all, :none, :new_episodes].each do |record_mode|
       context "for a :record => :#{record_mode} cassette with previously recorded interactions" do
-        subject { VCR::Cassette.new('example', :record => record_mode, :match_requests_on => [:uri]) }
+        let(:drop_unused_requests) { false }
```
Comment (olleolleolle) [inline spec/lib/vcr/cassette_spec.rb:614]:
```suggestion
        let(:drop_unused_requests) { false }

```

**Comment by olleolleolle** [comment]
(Ran the tests and recorded #948.)

**Review by olleolleolle — APPROVED** [review]
I like this change, and seems to do what it sets out to do.

(One thing we often ask is "How do we teach this?", and perhaps that's a further question, for later. But this is a new configuration option, and its place is in the user-visible documentation, in some form.)

The test failures in the cucumber tests have been reported separately.

**Comment by ryanseys** [comment]
Fix for curb issue here: https://github.com/vcr/vcr/pull/949

**Comment by ryanseys** [comment]
Depends on (and is rebased on) this PR, which we should merge first: https://github.com/vcr/vcr/pull/949

**Comment by ryanseys** [comment]
@olleolleolle 

> But this is a new configuration option, and its place is in the user-visible documentation, in some form.

I've added feature documentation in [`bf38ee2` (#946)](https://github.com/vcr/vcr/pull/946/commits/bf38ee2a4444b28c14dbf6ca3a152cc7f4f8517b)

**Comment by olleolleolle** [comment]
@ryanseys Cool, now with the other PR merged, we could re-run the tests, and that gave just one failure, which seems like a fixable one. Good job!

**Comment by ryanseys** [comment]
@olleolleolle Should be passing now. Please take another look.

**Comment by klyonrad** [comment]
@olleolleolle any chance for an official release of this feature? I tried it briefly and seemd quite awesome!

**Pull request merged**